### PR TITLE
jit: introduces Assembler interface towards homemade assembler

### PR DESCRIPTION
**Pull request by mathetake**
This introduces `Assembler` interface to abstract away `golang-asm`
behind it. Notably, new packages `asm`, `asm/amd64` and `asm/arm64`
are added under `internal/jit`. Now golang-asm is only used in three files
`golang_asm.go`, `amd64/golang_asm.go` and `arm64/golang_asm.go`
where we implement `amd64.Assembler` and `arm64.Assembler` interface.

part of #233 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Comment by mathetake** [comment]
OK passed on amd64 tomorrow i will do arm64 and then will get ready

**Comment by mathetake** [comment]
ok now compilable..

**Comment by mathetake** [comment]
ok cool passed all tests.. will polish and then ready..

**Comment by mathetake** [comment]
Desite the size of PR, there's nothing new, but rather code move and documentations!

File: internal/wasm/jit/asm/amd64/consts.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	ConditionalRegisterStateL                                                 // SF xor OF less (signed <)
+	ConditionalRegisterStateLE                                                // (SF xor OF) | ZF less or equal (signed <=)
+	ConditionalRegisterStateA                                                 // ˜CF & ˜ZF above (unsigned >)
+	ConditionalRegisterStateAE                                                // ˜CF above or equal (unsigned >=)
+	ConditionalRegisterStateB                                                 // CF below (unsigned <)
+	ConditionalRegisterStateBE                                                // CF | ZF below or equal (unsigned <=)
+)
+
+// Arm64-specific instructions.
+const (
```
Comment (mathetake) [inline internal/wasm/jit/asm/amd64/consts.go:22]: this is interesting list as this is the list of all instructions used in wazero!

File: internal/wasm/jit/asm/arm64/consts.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	REG_F29
+	REG_F30
+	REG_F31
+
+	// Floating point status register.
+	REG_FPSR
+)
+
+// Arm64-specific instructions.
+const (
```
Comment (mathetake) [inline internal/wasm/jit/asm/arm64/consts.go:102]: same for arm64!

File: internal/wasm/jit/jit_amd64.go [hunk 3]
```diff
@@ -4,21 +4,18 @@ package jit
...
-	"encoding/binary"
 	"fmt"
 	"math"
 	"runtime"
 	"unsafe"
 
-	asm "github.com/twitchyliquid64/golang-asm"
-	"github.com/twitchyliquid64/golang-asm/obj"
-	"github.com/twitchyliquid64/golang-asm/obj/x86"
-
```
Comment (mathetake) [inline internal/wasm/jit/jit_amd64.go:19]: no thirdpart deps in jit pkg! 👯

File: internal/wasm/jit/asm/amd64/golang_asm.go [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+	"fmt"
+
+	"github.com/twitchyliquid64/golang-asm/obj"
+	"github.com/twitchyliquid64/golang-asm/obj/x86"
+
+	"github.com/tetratelabs/wazero/internal/wasm/jit/asm"
+)
+
+// assemblerGoAsmImpl implements Assembler for golang-asm library.
+
```
Comment (codefromthecrypt) [inline internal/wasm/jit/asm/amd64/golang_asm.go:14]:
```suggestion
```

File: internal/wasm/jit/asm/golang_asm.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	"fmt"
+	"math"
+	"runtime"
+
+	goasm "github.com/twitchyliquid64/golang-asm"
+	"github.com/twitchyliquid64/golang-asm/obj"
+)
+
+// golangAsmNode implements Node for golang-asm library.
+
```
Comment (codefromthecrypt) [inline internal/wasm/jit/asm/golang_asm.go:14]:
```suggestion
```

**Review by codefromthecrypt — APPROVED** [review]
I love the organization. I'm sure you know the assemblers should have unit tests, but I also get that this is an interim step to keep the PR size down. Once you rip out twitchy, these would have tests.

**Comment by codefromthecrypt** [comment]
I also tested this manually with https://gist.github.com/codefromthecrypt/edb33284354d592dc6056b9b7263872f and the size didn't change and it also ran. Once twitchy is gone, I would expect the size to go down!

File: internal/wasm/jit/jit_amd64.go [hunk 6]
```diff
@@ -3723,26 +3413,26 @@ func (c *amd64Compiler) compileCallFunctionImpl(index wasm.Index, compiledFuncti
...
 	// here we get the actual offset in bytes via shifting callFrameStackPointerRegister by callFrameDataSizeMostSignificantSetBit.
 	// That is valid because the size of callFrame struct is a power of 2 (see TestVerifyOffsetValue), which means
 	// multiplying withe the size of struct equals shifting by its most significant bit.
-	c.compileConstToRegisterInstruction(x86.ASHLQ, int64(callFrameDataSizeMostSignificantSetBit), callFrameStackPointerRegister)
+	c.assembler.CompileConstToRegister(amd64.SHLQ, int64(callFrameDataSizeMostSignificantSetBit), callFrameStackPointerRegister)
 
 	// At this point, callFrameStackPointerRegister holds the offset in call frame slice in bytes,
 	// and tmpRegister holds the absolute address of the first item of call frame slice.
-	// To illustrate the situation:
+	// To illustrate the situation:m
```
Comment (nullpo-head) [inline internal/wasm/jit/jit_amd64.go:3435]:
```suggestion
	// To illustrate the situation:
```

**Pull request merged**